### ISSUE TITLE
Runtime's illink runs don't catch certain set of warnings

### ISSUE THREAD

**Issue by vitek-karas** [dotnet/runtime#84297]
If a framework library contains code which:
* Produces a trim warning
* Is not referenced from public API
* Is rooted via a LibraryBuild descriptor for testing purposes

Then runtime build will not report the trim warning anywhere. But dotnet/sdk will fail a test when such framework makes it into the SDK repo.

We have 3 runs of illink on a framework assembly:
* First is in runtime when we trim the assembly for shipping (the per assembly run). This runs with the "library" mode, but it also passes the `ILLink.Descriptors.LibraryBuild.xml` which is a descriptor, which roots more code (typically for testing purposes - I assume the tests create these via reflection for now). This run actually sees the problem in the code, but we disable all trim related warnings in this run, so it's not reported.
* Second is in runtime, the "sfx" run - where we trim the entire framework in "library" mode, but in this case, we don't apply the `ILLink.Descriptors.LibraryBuild.xml`, so the affected code is not rooted. And thus the trimmer doesn't even see the affected code.
* Third run is in SDK, where we trim the entire framework in "copy" mode, which sees all code, and thus reports the problem.

Found in https://github.com/dotnet/sdk/pull/31571. Related discussion in https://github.com/dotnet/runtime/pull/84272.

**Comment by vitek-karas** [comment]
@sbomer for ideas.

I think the cleanest solution would be to fix the "sfx" run to also merge all `LibraryBuild.xml` files and pass them to the linker. That would more closely mimic the per-assembly behavior.

**Comment by jkotas** [comment]
What the "sfx" run trying to test? If it is trying to test what happens on end-user machine, it should not use `LibraryBuild.xml` - `LibraryBuild.xml` is not going to be used on the end-user machine.

**Comment by vitek-karas** [comment]
> What the "sfx" run trying to test? If it is trying to test what happens on end-user machine, it should not use LibraryBuild.xml - LibraryBuild.xml is not going to be used on the end-user machine either.

It's there to report all of the trim warnings in the framework. It IS the run which should have reported this problem. (There's a matching "oob" run which does the same for all OOB assemblies).

While it's true that `LibaryBuild.xml` will not be there on user's machine, the code which it roots will be on the user machine. So if the user ends up marking the affected framework assembly as "copy", the trimmer would still see the affected code and report the problem to the user.

It's not very common that users will mark framework assemblies as copy, but it does happen (typically when they play the whack-a-mole game if trimming breaks their app and they don't care to fix the warnings).

That's basically what the test in SDK is trying to validate, that we're not shipping framework code which will produce trim warnings. The test originally targets down-level TFMs - meaning that latest SDK will not produce warnings when trimming downlevel apps. But we can't (shouldn't) disable the test for "current" TFM, since around RC that test needs to be passing for the current TFM, and we're bound to forget to enable it on-time.

**Comment by jkotas** [comment]
> So if the user ends up marking the affected framework assembly as "copy", the trimmer would still see the affected code and report the problem to the user.

Should the "sfx" run use the "copy" mode then?

**Comment by sbomer** [comment]
> The test did its job - it caught the problem. It just didn't do it before the code was merged into dotnet/runtime.

Agreed, so I'll close this issue. If this starts happening more often we can consider doing some extra validation in dotnet/runtime.
